### ISSUE TITLE
Discussion: LeaveOneGroupOut CrossValidation with caret

### ISSUE THREAD

**Issue by DeltaLima** [topepo/caret#602]
In Social psychology we are working on a project where we try to predict relationship quality (outcome) by personality (features). Main goal is to contribute to better match people with have higher chances to have a happy long lasting romantic relationship. I would be very grateful if you could help me with this by answering the following question:

At the moment, in R the k-fold-cv randomly sorts rows of data/people into the folds. A couple is represented by two rows in the dataset (partner 1 and partner 2) which are of course not always equally happy in the relationship they have with each other. But nevertheless the relationship quality of partner 1 and partner 2 correlate, which means the cases are somehow dependent. How can I sort partners of one couple to the same fold (but still as two cases), so that the test sample is always completely independent to the trainings sample? How can I write a Leave One Group Out CS - command in R, as it exists in Python (which I unfortunately cannot perform with)?

Couples are identified by the same number in the row paarID.

Here is the processing part of the code in R from the situation:

```
library(caret)
outcome <- "RQ_continuaryScale"
variables <- colnames(dat)[use_covar_i]
model <- paste(variables, collapse=" + ")
model <- paste(outcome, '~', model, collapse=' ')
training_config <- trainControl(method="cv", number=5, repeats = 100)
fit <- train(as.formula(model), data=dat_nomiss, "glmnet", trControl = training_config)
```

Thanks for your help!!

**Comment by m-dz** [comment]
Was just browsing through the issues looking for something else, but since I am here I think I can help. You could use the `trainControl()` `index` parameter to specify your own cv folds, now you just need to find a way to create those (e.g. sample from a couple ID?).

**Comment by topepo** [comment]
> How can I sort partners of one couple to the same fold (but still as two cases), so that the test sample is always completely independent to the trainings sample? 

I think that I get the issue but some mock data would help. Suppose we had
```r
ex_dat <- expand.grid(person1 = 1:5, person2 = 6:10)

set.seed(1)
ex_dat <- subset(ex_dat, runif(nrow(ex_dat)) < .4)
ex_dat <- ex_dat[order(ex_dat$person1, ex_dat$person2),]
rownames(ex_dat) <- 1:nrow(ex_dat)
```
```r
> ex_dat
   person1 person2
1        1       6
2        1       8
3        2       6
4        2       8
5        2      10
6        4       8
7        4       9
8        4      10
9        5       6
10       5       7
11       5      10
```
What is your independent experimental unit? Is it the pair or the person? I don't think that you have multiple rows on the pair (right?) so I would assume that it is the person that you want either in the training set or the test set. Please verify...

If that is the case, there is some new splitting code [see the section "Simple Splitting with Important Groups" in the docs](https://github.com/topepo/caret/blob/master/bookdown/03-Splitting.Rmd) slated for the next release (probably in late March) that you can use to split on the person and not the rows. 

If this is appropriate, you would then take the output of `groupKFold` and pass it into `trainControl`'s `index` argument (as you mentioned). 

Let us know if this is the right path...

**Comment by DeltaLima** [comment]
Hi! Thank you for your quick answer and sorry for the late response!

I've added some stripped down sample data to the attachments.
[Testdata_couples_1.csv 2.zip](https://github.com/topepo/caret/files/796416/Testdata_couples_1.csv.2.zip)

@topepo 

> What is your independent experimental unit? Is it the pair or the person? I don't think that you have multiple rows on the pair (right?) so I would assume that it is the person that you want either in the training set or the test set. Please verify...

A single person is our independent unit. They are indexed by "IDpaar" which are unique for a couple. So in my understanding, the data is already splitted. 

Here's the reading part of the code, for better understanding (is above the "processing part"):
`library(stats)
dat <- read.table('/Users/mhanisch/Downloads/Datensatz 2_192partner cross section3.csv', sep=';', dec=',', header=TRUE, stringsAsFactors=FALSE)
use_variables_i <- c(2, 4:63, 376:950, 1247:1414)
use_covar_i <- use_variables_i[2:length(use_variables_i)]
dat_nomiss <- dat[complete.cases(dat[, use_variables_i]),]`

@m-dz 

sounds interesting! that means that i have to read also the field IDpaar and use that as index for train control? 

> now you just need to find a way to create those (e.g. sample() from a couple ID or even use caret::createFolds() using such IDs?).

i think i haven't got it right now. i think i have to read bit about sample() and the createFolds()


Thank you so much for your support! :)

**Comment by DeltaLima** [comment]
@topepo 

Please forget my last answer.. I'm a noob to this topic. I had consultation with the study partner i work with on this topic:

One independent unit (test sample) should be a pair. A pair consists of two rows. One Partner in every row (https://github.com/topepo/caret/files/796416/Testdata_couples_1.csv.2.zip).
The name of the row which gives two partners of one couple the same number is named "PaarID"

One row has about 1000 columns (features) for one person.

I found following issue entry with code from you:

https://github.com/topepo/caret/issues/540
Is that what we need in this case? Unfortunately I don't know how to use it. :(

Thank you very much and sorry for the confusion with my post before!
best regards!

**Comment by topepo** [comment]
There is a little more documentation [here](https://github.com/topepo/caret/commit/6c684c52d42097dbf0460c20f0daf3b5965bbff2)
